mldsa/ntt.c

```diff
@@ -141,6 +143,44 @@ void ntt(int32_t a[MLDSA_N])
   /* are bounded in magnitude by 9 * MLDSA_Q                 */
 }
 
+/* Reference: Embedded into `invntt_tomont()` in the reference implementation
+ * [@REF] */
+static void mld_invntt_layer(int32_t r[MLDSA_N], unsigned layer)
+__contract__(
+  requires(memory_no_alias(r, sizeof(int32_t) * MLDSA_N))
+  requires(1 <= layer && layer <= 8)
+  requires(array_abs_bound(r, 0, MLDSA_N, MLDSA_Q))
+  assigns(memory_slice(r, sizeof(int32_t) * MLDSA_N))
+  ensures(array_abs_bound(r, 0, MLDSA_N, MLDSA_INVNTT_BOUND)))
```

Comment:
This won't work either.
The post-condition of the i-th layer needs to imply the pre-condition of the (i+1)-th layer.

Reply:
I think the pre-condition is
```
requires(array_abs_bound(r, 0, MLDSA_N, (MLDSA_N >> layer) * MLDSA_Q))
```
while the post-condition is:
```ensures(array_abs_bound(r, 0, MLDSA_N, (MLDSA_N >> (layer - 1)) * MLDSA_Q))
```
so that when layer=8, the bounds go from Q to 2Q. When layer=7, the bounds go from 2Q to 4Q and so on. When layer=1, the postconditions says that the final bound is 256Q.

The loop-invariant will need to be updated similarly.

Reply:
Sound about right.

Reply:
Addressed within [8049ec5](https://github.com/pq-code-package/mldsa-native/pull/179/commits/8049ec571905431dc8618c61e41c5ee31bfb60a5)